rocks/RocksIterator.java

```diff
@@ -42,35 +40,39 @@
         this.storage = storage;
         this.prefix = prefix;
         this.constructor = constructor;
-
-        isOpen = new AtomicBoolean(true);
         state = State.INIT;
     }
 
-    private void initalise() {
-        this.internalRocksIterator = storage.getInternalRocksIterator();
-        this.internalRocksIterator.seek(prefix);
-    }
-
-    private boolean fetchAndCheck() {
-        byte[] key;
-        if (!internalRocksIterator.isValid() || !bytesHavePrefix(key = internalRocksIterator.key(), prefix)) {
-            state = State.COMPLETED;
-            recycle();
+    private synchronized boolean initialise() {
+        if (state != State.COMPLETED) {
+            this.internalRocksIterator = storage.getInternalRocksIterator();
+            this.internalRocksIterator.seek(prefix);
+            state = State.EMPTY;
+            byte[] key;
+            if (!internalRocksIterator.isValid() || !bytesHavePrefix(key = internalRocksIterator.key(), prefix)) {
+                recycle();
+                return false;
+            }
+            next = constructor.apply(key, internalRocksIterator.value());
+            state = State.FETCHED;
+            return true;
+        } else {
```

Comment:
Why are duplicating the code from `fetchAndCheck()` into `initialise()`, @flyingsilverfin ? I don't think this is a good idea.

Reply:
`initialise()` does more than just "initialising" now.

Reply:
renamed and refactored